### ISSUE TITLE
Bold plan for a major release

### ISSUE THREAD

**Issue by gummif** [zeromq/cppzmq#400]
Here are a few ideas for changes to be made for (potentially breaking) cppzmq v5.

1. Require C++11 standard support, but still support gcc 4.7 and later with partial C++11 support. This possible future requirement has been in the read me for 2 years now. Large fraction of the library now requires C++11.
2. Require libzmq 4.0+. There are no CI tests with older versions and there are no releases on github for version prior to 4.0.6. (should we maybe require 4.1 or 4.2?)
3. Remove deprecated functions that would not cause silent breakage (would cause compile time errors). Keep functions that would cause silent changes to runtime behavior if removed (e.g. greedy function overloads).

These changes are conservative in the sense that if you are using C++11 or later, libzmq v4 or later and not using any deprecated functionality, then the upgrade from the latest v4 to v5 should not require any changes to you code. If you do use deprecated functions then you might get a compile time error in the worst case.

**Comment by bluca** [comment]
Please do not break backward compatibility. This header is shipped in distributions. It would cause major, major headaches and work to downstreams.

**Comment by gummif** [comment]
After some deep though I can see that part 3 and maybe part 2 could cause some pain with system packages (system packages are generally a bad idea anyway for development). I don't know how the process is for adding new package versions to distributions, but I guess it is their responsibility to not add package versions that would break the system? But if it is possible people will eventually managed to break things.

About requiring C++11, the read me says "Some features already require a compiler supporting C++11. In the future, probably all features will require C++11". Is that something that you prefer never happens?

And regarding part2, I believe libzmq v3.x is rarely used. And it would be very unlikely that someone would use libzmq v3.x together with latest cppzmq, so in practice that would requirement should have no effect (in that event an option would be to simply update libzmq or not update cppzmq). Correct me if you can come up with a situation where that would be the case.

**Comment by bluca** [comment]
> I don't know how the process is for adding new package versions to distributions, but I guess it is their responsibility to not add package versions that would break the system?

Yes, but it means there will never be a new version uploaded, which is not good.

The problem is not C++11 so much for distros - they all support it, it's just the ancient and deprecated stuff that doesn't. Not even zmq 3.x. The problem is breaking API compatibility by removing functions.

**Comment by bluca** [comment]
That would be fine, yes

**Comment by planetmarshall** [comment]
> Here are a few ideas for changes to be made for (potentially breaking) cppzmq v5.
> 
> 1. Require C++11 standard support, but still support gcc 4.7 and later with partial C++11 support. This possible future requirement has been in the read me for 2 years now. Large fraction of the library now requires C++11.

Skip C++11, C++14 and C++17, move to C++20 and provide an asynchronous API based on coroutines, mirroring asyncio support in the PyZMQ library.

**Comment by gummif** [comment]
I think coroutine support could be added as an addon for those that have compilers that support it. I have been thinking about it but have not yet thought of a good solution (adding an async API on top of a blocking API is not ideal).

**Comment by SidneyCogdill** [comment]
I've wrote two helper functions for Boost.Cobalt to asynchronously receive messages (works on Linux for me, probably on BSD / macOS too):

```cpp
namespace cobalt = boost::cobalt;
namespace asio = boost::asio;
namespace scope = boost::scope;

auto async_recv(zmq::socket_t &socket, zmq::message_t &message) -> cobalt::task<zmq::recv_result_t>
{
    asio::posix::stream_descriptor sd{co_await cobalt::this_coro::executor,
                                      socket.get(zmq::sockopt::fd)};
    /// Since stream_descriptor takes the ownership of the file descriptor,
    /// it will attempt to close the descriptor on scope exit.
    /// This defer guard prevents that from happening,
    /// by releasing the fd right before exiting the scope.
    scope::defer_guard release_fd = [&] { sd.release(); };

    while (true) {
        auto events = socket.get(zmq::sockopt::events);
        if (events & ZMQ_POLLIN)
            co_return socket.recv(message, zmq::recv_flags::dontwait);
        co_await sd.async_wait(asio::posix::descriptor_base::wait_read);
    }
}

auto async_recv(zmq::socket_t &socket, zmq::multipart_t &multipart) -> cobalt::task<bool>
{
    multipart.clear();
    bool more = true;
    /// The control flow is directly copied from the member function `zmq::multipart_t::recv`,
    /// but with the recv replaced by the async version.
    while (more) {
        zmq::message_t message;
        if (!co_await async_recv(socket, message))
            co_return false;
        more = message.more();
        multipart.add(std::move(message));
    }
    co_return true;
}
```

It can be used like this:

```cpp
zmq::socket_t socket{ctx, zmq::socket_type::rep};
socket.bind("tcp://127.0.0.1:5555");
zmq::multipart_t a;
co_await async_recv(socket, a);
std::println("Received {} messages", a.size());
```

On `co_await`, the current coroutine is suspended and any other tasks can continue execute concurrently.

Not sure how to make it work on Windows *yet* since I don't have access to Windows machine right away.

**Comment by SidneyCogdill** [comment]
An official coroutine support would be great (albeit hard, as zmq doesn't have callbacks like other pre-C++20 asynchronous libraries, and C++20 coroutines are basically encapsulation of callbacks). the helper functions for cobalt inherently relies on asio to watch the file descriptor. I don't think it will work for any other protocol than TCP.

**Comment by SidneyCogdill** [comment]
FYI I've implemented Sender & Receiver support (over 1000 LoC). I haven't wrote test cases for it yet. Unlike the corral coroutine one, this is thread safe (at the cost of some `atomic<bool>` fetching). It still depends on Asio to listen on the file descriptor events. It also depends on [stdexec](https://github.com/NVIDIA/stdexec) (but this will change once C++26 standard library is implemented by major vendors).

There is a performance hazard which I haven't decided the solution yet: When many actors are attempting to send/receive multipart messages on the same socket, with the current implementation every operation is `async_wait`-ed and will cause excessive wakeup. In other words it doesn't scale linearly with the amount of pending multipart message operations.

I can address this by adding a callback queue and only awake one operation per file descriptor event, which will make the performance scale linearly. However this also makes all multipart operations perform slightly slower as there is now the additional cost of the callback queue. More importantly, operating multipart message simultaneously from multiple actors is just not a good idea in general as it's not designed to be thread safe, and I don't know if it's worth adding that overhead to improve performance for the case you want to avoid if you actually care about performance.

**Comment by SidneyCogdill** [comment]
Another hazard is when both the `recv` (single message) and `recv_multipart` are executing in parallel, theoretically there is a small window for race condition where `recv` enters the critical section but `recv_multipart` isn't aware of that. Then `recv` might accidentally take a message part from the multipart message, breaking the atomic property of multipart message. This can be trivially mitigated by making `recv` also set the socket-wide critical section flag, however again at the cost of slowing down all `recv` operations (2 exchange + 1 store + 1 load for each `recv`, where it originally only requires 1 load + 1 exchange).
